Feat - Core Location 구현하기

longlivedrgn opened the pull request:
안녕하세요 써니 testable한 core location을 만들어봤습니다~

> 추후에 실제로 Core Location을 다른 Model과 연결할 때는 코드가 조금 달라질 거 같긴한데, 그 부분은 실제 모델과 연결할 때 고민해봐도 좋을 거 같아요.(+얼럿을 통하여 사용자의 위치 정보 수집 승인과 관련된 test는 아직 하지 않았어요..ㅎㅎ)

- 사실 testable하게 만들지 않으면 훨씬 더 직관적이고 간단하게 코드를 구현할 수 있을 거 같은데, 테스트를 하려고 하다보니, `CLLocationManagerDelegate`과의 의존성을 끊어야되고, 새로운 protocol(LocationFetcher)을 정의해서 의존성 주입을 해주어야했어요.

- 그래서, 저희가 WeatherForecast에서 활용한 Core Location 코드와 현재 코드를 비교하면서 제가 어떻게 Core Location을 구현했는 지 이야기해볼게요.

저희는 이렇게 Core Location Manager 코드를 구현했었어요.
- CLLocationManager 객체를 갖고
- CLLocationManagerDelegate을 채택해서, 에러 처리 및 사용자 위치 정보 승인 요구, 위치 정보 업데이트를 해요.
```swift
final class CoreLocationManager: NSObject {
    
    weak var delegate: CoreLocationManagerDelegate?
    private var locationManager = CLLocationManager()
    
    override init() {
        super.init()
        locationManager.desiredAccuracy = kCLLocationAccuracyKilometer
        locationManager.requestWhenInUseAuthorization()
        locationManager.delegate = self
    }
    
    func changeGeocoder(location: CLLocation) async throws -> CLPlacemark? {
        
        let geocoder = CLGeocoder()
        let locale = Locale(identifier: "Ko-kr")
        let locations = try await geocoder.reverseGeocodeLocation(location, preferredLocale: locale)
        let firstLocation = locations.last
        
        return firstLocation
    }
}

extension CoreLocationManager: CLLocationManagerDelegate {
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        
        guard let location = locations.last else { return }
        delegate?.coreLocationManager(self, didUpdateLocation: location)
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print("locationManager-didFailWithError")
    }
    
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        
        switch manager.authorizationStatus {
        case .authorizedWhenInUse:
            manager.requestLocation()
            
        case .denied, .restricted:
            break
            
        case .notDetermined:
            manager.requestWhenInUseAuthorization()
            
        default:
            return
        }
    }
}
```

- 그리고, 제가 구현한 Core Location은 아래와 같아요.
```swift
final class LocationManager: NSObject {

    typealias LocationCallBack = ((CLLocation?, Error?) -> Void)

    private var locationFetcher: LocationFetcher
    var locationCallBack: LocationCallBack?

    init(locationFetcher: LocationFetcher = CLLocationManager()) {
        self.locationFetcher = locationFetcher
        super.init()
        self.locationFetcher.locationFetcherDelegate = self
    }

    func fetchCurrentLocation(completion: @escaping LocationCallBack) {
        self.locationCallBack = completion
    }

    func updateLocation() {
        locationFetcher.startUpdatingLocation()
    }

}
```

- WetherForcast와 GoodMorning 서로 비슷한 역할을 하는 객체는 아래와 같이 연결할 수 있어요.

CoreLocationManager - LocationManager
CLLocationManager() - LocationFetcher
CLLocationManagerDelegate - LocationFetcherDelegate

- CLLocationManager()의 역할을 protocol로 만든 것이 LocationFetcher이고, CLLocationManagerDelegate의 역할을 protocol로 만든 것이 LocationFetcherDelegate이에요.

- 그래서 LocationFetcher의 메소드를 보면 실제 CLLocationManager()가 하는 역할들을 정의해놓고 있어요.
```swift
protocol LocationFetcher {

    var locationFetcherDelegate: LocationFetcherDelegate? {get set}

    func startUpdatingLocation()
    func requestWhenInUseAuthorization()
    func requestAlwaysAuthorization()

}
```

- LocationFetcherDelegate도 실제 CLLocationManagerDelegate의 메소드와 비슷한 모습의 메소드를 가지고 있음을 알 수 있어요.
```swift
protocol LocationFetcherDelegate: AnyObject {

    func locationFetcher(_ fetcher: LocationFetcher, didUpdateLocations locations: [CLLocation])
    func locationFetcher(_ fetcher: LocationFetcher, didFailWithError error: Error)
    func locationFetcher(
        _ fetcher: LocationFetcher,
        didChangeAuthorization authorization: CLAuthorizationStatus
    )

}
```

- 그래서, 실제 CLLocationManagerDelegate의 메소드는 다시 LocationFetcherDelegate의 메소드를 호출해요.
```swift
extension LocationManager: CLLocationManagerDelegate {

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        locationFetcher(manager, didUpdateLocations: locations)
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        locationFetcher(manager, didFailWithError: error)
    }

    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        locationFetcher(manager, didChangeAuthorization: manager.authorizationStatus)
    }

}
```

- 그리고 LocationFetcherDelegate의 메소드는 LocationManager의 closure에 받아온 위치를 담아둬요.
```swift
    func locationFetcher(_ fetcher: LocationFetcher, didUpdateLocations locations: [CLLocation]) {
        if let location = locations.first {
            locationCallBack?(location, nil)
        } else {
            locationCallBack?(nil, LocationError.canNotBeLocated)
        }

    }
```

- 그리고 실제 위치를 받아올 때는 아래와 같이 completionHandler를 통해서 위치 정보를 가져와요.
```swift
🌟LocationManager
final class LocationManager: NSObject {

    typealias LocationCallBack = ((CLLocation?, Error?) -> Void)

    private var locationFetcher: LocationFetcher
    var locationCallBack: LocationCallBack?

...

    func fetchCurrentLocation(completion: @escaping LocationCallBack) {
        self.locationCallBack = completion
    }
}

🌟 View Controller
class ViewController: UIViewController {

    private let locationManger = LocationManager()

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .red
        requestCurrentLocation()
    }

    func requestCurrentLocation() {
        locationManger.fetchCurrentLocation { location, error in
            if let error {
                print(error)
            } else {
                print(location)
            }
        }
    }

}

```

- 테스트는 MockLocationFetcher를 정의해서 내가 임의의로 넣어준 위치정보를 제대로 받아오는 지와 에러를 제대로 던지는 지를 확인해봤어요. (이 부분과 관련되어서 궁금한 거 있으면 comment 달아주세요)

참고한 Core Location 문서
- https://developer.apple.com/documentation/corelocation
- https://developer.apple.com/documentation/corelocation/configuring_your_app_to_use_location_services
- https://developer.apple.com/videos/play/wwdc2018/417/
- https://longlivedrgn-miro.tistory.com/61

longlivedrgn committed:
feat: LocationManager 로직 구현

longlivedrgn committed:
feat: LocationFetcher 및 LcoationFetcherDelegate 프로토콜 구현

longlivedrgn committed:
chore: LocationCallBack(completion block) 활용하도록 리팩토링

longlivedrgn committed:
feat: View Controller에서 Core Location API 호출

longlivedrgn committed:
test: fetchCurrentLocation 함수가 제대로 작동하는 지 확인하는 테스트 코드 추가

longlivedrgn committed:
chore: Core Location test 폴더링

longlivedrgn committed:
test: Core Location error 테스트 코드 작성

longlivedrgn committed:
test: Core Location error 테스트 코드 작성

longlivedrgn committed:
Merge branch 'feat/2/Core-Location' of https://github.com/longlivedrgn/GoodMorning into feat/2/Core-Location

SunnnySong commented on GoodMorning/GoodMorning/CoreLocation/Interface/LocationFetcher.swift:
```diff
@@ -0,0 +1,18 @@
+
+import CoreLocation
+
+protocol LocationFetcher {
```
Swift의 일반적인 컨벤션에 따라 -able 네이밍은 어떨까요?
LocationFetchable을 채택한 객체의 네이밍이 LocationFetcher가 되는 것이 조금 더 자연스럽지 않을까? 하는 생각이 들었어요!

longlivedrgn replied:
넵! 로케이션을 가져올 수 있는 이라는 LocationFetchable 네이밍으로 변경하겠습니다!

SunnnySong commented on GoodMorning/GoodMorningTests/CoreLocationTest/CoreLocationTest.swift:
```diff
@@ -0,0 +1,56 @@
+import CoreLocation
+@testable import GoodMorning
+
+final class CoreLocationTest: XCTestCase {
```
미로의 CoreLocation Unit Test를 보고 많은 것을 배웠습니다😃 나중에 관련 내용을 기억하기 위해 잠시 정리 한번 할게요!

CLLocation을 사용해 위치 데이터를 가져올 땐 권한 설정과 실제 기기/시뮬레이터의 존재, 두 가지 제약이 존재합니다. 
Unit Test를 진행할 때에는 두 가지 제약 조건으로부터 벗어나 위치 데이터를 가져오고 테스트 할 수 있어야 합니다.

longlivedrgn replied:
넵! 두 가지의 의존성을 제거하기 위해서 Mock을 만들었습니다 :)

SunnnySong commented on GoodMorning/GoodMorningTests/CoreLocationTest/CoreLocationTest.swift:
```diff
@@ -0,0 +1,56 @@
+        locationFetcher = nil
+    }
+
+    func test_임의의_위치를_넣어줬을때_제대로_위치_정보를_가져오는지_확인() {
+        let expectationLocation = CLLocation(latitude: 10, longitude: 10)
+        var resultLocation: CLLocation?
+
+        sut.fetchCurrentLocation { location, error in
+            resultLocation = location
+        }
+        locationFetcher.locationCallBackValue = CLLocation(latitude: 10, longitude: 10)
+        locationFetcher.locationErrorCallBackValue = nil
+        sut.updateLocation()
+
+        XCTAssertEqual(resultLocation?.coordinate.latitude, expectationLocation.coordinate.latitude)
+    }
```
CLLocation 에 관련된 Unit Test를 보고 궁금했었던 점이 있어요!

해당 test는 `expectationLocation`라는 임의의 위치 데이터가 존재하고, fetchCurrentLocation에 들어오는 데이터와 동일한지 확인하고 있습니다. 

그럼 이 테스트의 목적은 LocationManager와 LocationFetcher로 분리한 CLLocation 관련 객체들이 원래의 의도대로 맞게 작동하고 있는지에 대한 여부를 판단하는 테스트일까요?

longlivedrgn replied:
앗 맞습니다!

해당 테스트를 하는 목적은 LocationManager 타입의 메소드 fetchCurrentLocation 메소드가 제대로 작동하는 지를 판단하는 테스트입니다!

LocationManager는 CLLocationManager를 갖고, 해당 타입을 통해서 location을 받는데 LocationManager의 fetchCurrentLocation 메소드를 테스트하기 위해서는 CLLocationManager에 의존을 해야합니다. 

그러나 CLLocationManager는 위에서 말씀해주신대로 2가지 의존성을 가지고 있기에 CLLocationManager 역할을 하는 LocationFetchable 타입을 채택한 MockLocationFetcher를 생성하여 메소드를 테스트 하는 것입니다!

이 과정에서 CLLocationManager가 location 가져오는 작업(Delegate 메소드 호출)을 MockLocationFetcher에서는 location을 외부에서 프로퍼티로 설정을 할 수 있도록 해주었습니다!

SunnnySong commented on GoodMorning/GoodMorningTests/CoreLocationTest/MockLocationFetcher.swift:
```diff
@@ -0,0 +1,30 @@
+import CoreLocation
+@testable import GoodMorning
+
+class MockLocationFetcher: LocationFetcher {
```
MockLocationFetcher이라는 타입을 생성해 Unit Test를 진행하는 아이디어가 참 좋은 것 같습니다👍🏻

longlivedrgn replied:
💪

SunnnySong commented on GoodMorning/GoodMorning/ViewController.swift:
```diff
@@ -6,11 +6,26 @@
+        requestCurrentLocation()
+    }
+
+    func requestCurrentLocation() {
+        locationManger.fetchCurrentLocation { location, error in
+            if let error {
+                print(error)
+            } else {
+                print(location)
+            }
+        }
```
해당 코드는 `locationManager`로 위치 데이터가 잘 들어오는지 테스트 하기 위한 코드로 develop branch로 머지 되기 전에 삭제하는 것이 좋을 것 같아요~

longlivedrgn replied:
넵! 써니가 코드 리뷰하실 때 확인하기 좋게 만들었는데, 테스트 코드로 작성하는 게 더 좋았을 수도 있겠네요!

SunnnySong commented on GoodMorning/GoodMorning/CoreLocation/LocationError.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+enum LocationError: Error {
+
+    case canNotBeLocated
+
+}
```
CoreLocation 폴더 하위에 Error 폴더 생성 후 LocationError를 넣으면 좋을 것 같습니다~ 
아니면 제일 상위 폴더에서 Error라는 폴더 하나 생성 후 모든 Error 타입들을 한꺼번에 관리할까요?

longlivedrgn replied:
앗 폴더링을 안했네요!ㅎㅎ 감사합니다 ~!

SunnnySong commented on GoodMorning/GoodMorning/CoreLocation/Interface/LocationFetcherDelegate.swift:
```diff
@@ -0,0 +1,19 @@
+//  Created by Miro on 2023/07/17.
+//
+
+import CoreLocation
```
CLLocationManagerDelegate함수에서 LocationFetcherDelegate를 호출하는 부분에서 CLLocation을 Coordinator로 변환해서 넘겨주면 어떨까 생각했어요😀

그럼 해당 파일에 `import CoreLocation`를 할 필요도 없고, 실제 위치 데이터를 전달해줄 때, CLLocation이 아니라 Domain Model인 Coordinator을 사용할 수 있으니 변환 과정이 생략 가능하지 않을까 해요!

SunnnySong commented on GoodMorning/GoodMorning/CoreLocation/Extensions/CLLocationManager+.swift:
```diff
@@ -0,0 +1,21 @@
+
+import CoreLocation
+
+extension CLLocationManager: LocationFetcher {
```
저는 Core Location 코드 중에서 해당 코드가 가장 신기했습니다 ㅋㅋㅋ

longlivedrgn committed:
chore: LocationFecher 프로토콜 LocationFetchabel로 네이밍 변경

longlivedrgn committed:
chore: ViewController내에 불필요한 코드 삭제

longlivedrgn committed:
chore: Coordinate 타입 정의 및 연관 메소드 수정

- 기존의 CLLocation을 return 하는 locationfetcherdelegate 메소드를 Coordinate 타입을 return하게 변경

SunnnySong commented:
MergeDevelop 으로 머지 완료~
